### PR TITLE
ci test: build, run and test kong-operator

### LINKED ISSUE #20: CI: Test the "vanilla" operator at HEAD
Implement a GitHub Actions workflow which:
- spins up a Kubernetes cluster (e.g. `microk8s` with `dns`, `rbac`, `registry` addons)
- builds the operator from HEAD and pushes to the local registry
- runs the operator from the image built in the previous step
- deploys an example `Kong`
- waits until Kong is up
- deploys an example `Ingress`
- makes an HTTP request to validate that the ingress controller handled the request properly

### PR DESCRIPTION
**Pull request by mflendrich**
closes #20 

This PR:
- defines a GitHub Actions workflow which:
    - spins up `microk8s` with a local image registry
    - builds the operator image and pushes it to the local registry
    - runs the operator and instructs it to spin up two Kongs watching different ingress classes
    - creates two ingresses in different ingress classes
    - validates that the right Kong supports the right Ingress and not the other way around
- slightly reorganizes `deploy/` in order to maintain its current capability (of spinning up kong-operator) as well as allowing to start kong-operator from an image in the local registry (this could have been achieved using e.g. kustomize but I opted for a simpler approach)

The test takes about 3 minutes to succeed (and about 5 to fail). I've tried to contain all possible races by implementing `wait`s in the test, so the most likely source of potential flakiness is slow (timing out) downloads.

I've also implemented a helper (to wait on arbitrary jsonpath in k8s objects) that can be superseded by kubernetes/kubernetes#83094.

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    steps:
+    - name: 'Check out the repo'
+      uses: actions/checkout@v2
+
+    - name: 'Arrange: Set up microk8s'
+      run: |
+        sudo snap install microk8s --channel 1.18/stable --classic
+        for addon in dns rbac registry
+        do
+          sudo microk8s enable $addon
```
Comment (hutchic) [inline .github/workflows/test.yaml:17]: the situation may have changed but FWIW in the past weused microk8s and migrated to kind because of challenges with microk8s in CI

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    - name: 'Arrange: Set up Kong Operator'
+      run: |
+        . test/integration/helpers.sh
+
+        kubectl apply -f deploy/crds/charts_v1alpha1_kong_crd.yaml
+        kubectl apply -f deploy/
+        kubectl apply -f test/integration/operator-deployment.yaml
+
+        wait_for 'kubectl get deploy -n kong kong-operator -o name' ".+" 30
+        kubectl wait --for=condition=Available deploy -n kong kong-operator --timeout=120s
```
Comment (hutchic) [inline .github/workflows/test.yaml:41]: TIL

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        wait_for 'kubectl get deploy kong1-kong -o name' ".+" 30
+        kubectl wait --for=condition=Available deploy kong1-kong --timeout=120s
+        wait_for 'kubectl get deploy kong2-kong -o name' ".+" 30
+        kubectl wait --for=condition=Available deploy kong2-kong --timeout=120s
+
+        kubectl wait --for=condition=Available deploy echo --timeout=120s
+        wait_for 'kubectl get ingress demo1 -o jsonpath={.status.loadBalancer.ingress}' ".+" 120
+        wait_for 'kubectl get ingress demo2 -o jsonpath={.status.loadBalancer.ingress}' ".+" 120
+
+    - name: 'Assert: Validate Ingress'
```
Comment (hutchic) [inline .github/workflows/test.yaml:65]: if you want to recycle some prior art kong-build-tools uses tavern testing to do this sort of testing

https://github.com/Kong/kong-build-tools/tree/master/test/tests/02-api

We've done exactly this for testing `docker-kong` here: https://github.com/Kong/docker-kong/blob/master/tests.sh#L85-L88

TLDR
```
git clone kong-build-tools@tag
export RESTY_VERSION KONG_VERSION KONG_ADMIN_URI KONG_PROXY_URI
pushd kong-build-tools
make test
```

At this time it's probably overkill as this is ran on PR/commits to kong, kong releases and PR/commits to docker-kong. I can't think of anyway a bug this would have caught slips through all those spots and would be miraciously caught by running the identical test suite here.

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        set -x
+
+        PROXY1_IP=$(kubectl get service kong1-kong-proxy -o jsonpath={.spec.clusterIP})
+        PROXY2_IP=$(kubectl get service kong2-kong-proxy -o jsonpath={.spec.clusterIP})
+
+        [ "$(curl -sw '%{http_code}' -o /dev/null http://$PROXY1_IP/via-kong1/)" == 200 ]
+        [ "$(curl -sw '%{http_code}' -o /dev/null http://$PROXY2_IP/via-kong2/)" == 200 ]
+
+        [ "$(curl -sw '%{http_code}' -o /dev/null http://$PROXY1_IP/via-kong2/)" == 404 ]
+        [ "$(curl -sw '%{http_code}' -o /dev/null http://$PROXY2_IP/via-kong1/)" == 404 ]
```
Comment (hutchic) [inline .github/workflows/test.yaml:76]: can you do some basic "what went wrong" if these fail for these commands?

logs, kubectl status etc etc

File: .github/workflows/test.yaml (hunk 3 above)
Comment (mflendrich) [reply .github/workflows/test.yaml:65]: It's my guess that the test suite, which you're referring to, validates a single instance of Kong **gateway** which imo is out of scope for this test.

This test tells Kong **Operator** to spin up two instances of Kong **ingress controller**, and then validates that the ICs are up and that configuration doesn't leak from one IC to the other. FWIW we wouldn't be able to achieve that with the tests from kong-build-tools.

File: .github/workflows/test.yaml (hunk 1 above)
Comment (mflendrich) [reply .github/workflows/test.yaml:17]: Interesting. What were the challenges?

File: .github/workflows/test.yaml (hunk 4 above)
Comment (mflendrich) [reply .github/workflows/test.yaml:76]: Done. PTAL at added commits

File: .github/workflows/test.yaml (hunk 1 above)
Comment (hutchic) [reply .github/workflows/test.yaml:17]: swapping versions so we could test a matrix of k8s versions was challenging at times.

permissions required to install it and it's dependencies including on dev machines which are a mix of nix and MacOS mostly. kind mostly just ran the same on CI as it did locally.

Our situation was somewhat different though the team isn't versed in k8s so there's that

File: .github/workflows/test.yaml [hunk 5]
```diff
@@ -0,0 +1,91 @@
+name: Test
+
+on: [push]
```
Comment (hutchic) [inline .github/workflows/test.yaml:3]: your situation might be unique however, we've been finding pr's to master and push's to master. The logic being we care about what folks want to merge into master, we care about master itself, the tip of a development branch doesn't matter.

Keeping in mind we do this to save money on CI we host though maybe just leave this as run on all?

Comment (rainest) [reply .github/workflows/test.yaml:3]: On PR is typical for our team's other repos (I think--it may be on other pushes, I just don't check until I PR stuff)
